[1단계 - 로또(자동)] 써머(최혜원) 미션 제출합니다

hyewoncc opened the pull request:
안녕하세요, 철시! 😆 
로또 미션 리뷰 매칭된 써머입니다, 잘부탁드려요

이번 미션은 `TDD, 원시값 포장, 일급 컬렉션, enum 사용` 을 중점으로 진행했는데요,  
질문 할 사항은 파일에 달아두도록 하겠습니다!

Seongwon97 committed:
docs: 요구사항 정리

Seongwon97 committed:
feat: 로또 넘버 객체 생성

- 같은 번호를 가진 로또 넘버 객체는 같은 객체라고 인지한다.
- readme에 요구사항 추가

Seongwon97 committed:
feat: 로또 넘버 객체 리스트 일급 컬렉션 구현

- 매개변수로 받은 Integer리스트를 옳바르게 저장하는지 테스트 완료

Seongwon97 committed:
feat: 당첨 등수 판별 기능 구현

- 당첨 등수 판별 기능 구현
- 상금 정보를 담은 enum생성

Seongwon97 committed:
feat: 로또 결과 도출 메서드 구현

- 로또 결과를 Map으로 도출하는 메서드 구현
- 전략 패턴을 사용하여 랜덤하게 로또 번호를 생성해주는 메서드 구현

Seongwon97 committed:
feat: 수익률 계산 메서드 구현 완료

- 구매 금액과 상금을 비교하여 수익률을 계산하는 메서드 구현 완료

Seongwon97 committed:
style: formatting 통일

Seongwon97 committed:
refactor: 전략 패키지 파일 분리

- 전략 패키지 관련 파일을 /domain/strategy로 분리

Seongwon97 committed:
refactor: 최종 결과 객체를 일급 컬렉션으로 분리

- PurchadedLotto의 Map객체를 일급 컬렉션으로 분리

Seongwon97 committed:
style: 불필요한 import제거

Seongwon97 committed:
feat: 가격 입력값이 정수가 아닌 경우 예외처리

Seongwon97 committed:
feat: 가격 입력값이 음수인 경우 예외처리

Seongwon97 committed:
refactor: 가격 입력이 1000원 미만일 떄 예외가 발생하도록 수정

- 음의 입력값만 체크하던 메서드를 최소 금액보다 낮을 떄 예외가 발생하도록 수정

Seongwon97 committed:
feat: 6개가 아닌 로또 넘버 입력시 예외처리

Seongwon97 committed:
feat: 로또 번호가 숫자가 아닌 경우 예외처리

Seongwon97 committed:
feat: 로또 번호가 1~45 사이가 아닐 경우 예외처리

Seongwon97 committed:
feat: 중복된 로또 번호 존재시 예외처리

Seongwon97 committed:
feat: 보너스 번호를 이미 갖고 있을 시 예외처리

- 당첨 번호 객체 생성
- 보너스번호와 일반 번호와 중복시 예외처리

Seongwon97 committed:
feat: 로또 구매 기능 구현 완료

- View와 Model의 메서드를 호출하여 연결 완료

Seongwon97 committed:
feat: 당첨번호, 보너스 번호 생성 구현 완료

- 보너스 번호 중복시 반복 입력 구현

Seongwon97 committed:
fix: 3개 미만 일치시 발생하는 오류 수정

Seongwon97 committed:
feat: 최종 상금 통계 출력 기능 구현

Seongwon97 committed:
fix: 최종 상금 통계 출력 순서 수정

- 최종 상금 출력을 5등부터 1등까지 순차적으로 출력되도록 수정

Seongwon97 committed:
feat: 최종 수익률 출력 기능 구현

Seongwon97 committed:
refacter: 매직넘버를 상수로 변경, 불필요한 import제거

- 파일들에 있는 매직넘버들을 모두 상수로 변경
- 불필요한 import들을 제거
- 파일들의 Indent를 통일

Seongwon97 committed:
refactor: 우승 상금 Enum클래스 내부 로직 변경

- WinnerPrice.java -> Prize.java로 이름 변경
- Enum 필드에 bonus정보 추가
- 등수 찾는 로직 개선

Seongwon97 committed:
refactor: 변수, 메서드 명을 성격에 맞게 상금을 prize로 변경

Seongwon97 committed:
refactor: 테스트의 displayName설정

- InputValidationTest.java의 테스트들을 nested 테스트로 묶음
- 테스트 클래스들에 DisplayName추가

Seongwon97 committed:
style: 컨벤션 통일

hyewoncc commented on src/main/java/controller/LottoGameController.java:
```diff
@@ -0,0 +1,36 @@
+        purchasedLotto = new PurchasedLotto(InputView.inputPrice());
+        OutputView.printPurchasedLotto(purchasedLotto);
+        initWinningNumber();
+        purchasedLotto.calculatePrizeResult(winningNumber);
```
인스턴스 필드를 제거하고 메서드의 반환형으로 받는 게 나을까요? 
컨트롤러가 두 객체 관리하는 책임을 진다고 생각해서 필드로 만들었는데,  
객체의 관리(=책임)는 객체 스스로가 셀프로 하고, 컨트롤러는 관리가 아닌 메서드 실행 제어만 하는게 맞는지 고민이 됩니다

pkch93 replied:
컨트롤러에 역할에 대해 고민중이시군요. 컨트롤러는 보통 상태를 가지지 않고 실행 제어의 역할을 하죠. 이에 대해 한번 고민해보면 좋을거 같아요. 혹시 고민의 결과가 있다면 코멘트에 남겨주세요 :)

hyewoncc replied:
리팩토링 하며 생각해봤는데, 상태를 갖고 관리하는 순간 컨트롤러가 아닌 모델에 가까워지는 것 같습니다
모델이 로직에 필요한 모든 상태를 나눠 갖고, 그 상태를 메세지 형태로 꺼내서 컨트롤러를 통해 주고받는 것이 객체지향적인 동시에 mvc 패턴에 부합하는 것 같아요

hyewoncc commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -0,0 +1,37 @@
+
+    public LottoNumber(int number) {
+        this.number = number;
+    }
```
pr을 올리고 나서 원시값 포장에 대한 글을 읽어봤는데,
현재 view에서 1-45가 넘는 값이 입력 되었을 때 예외를 발생시킵니다
그런데 LottoNumber 스스로가 1-45 이 외의 숫자가 생성 파라미터로 들어왔을 때 예외를 발생시키는 것이
LottoNumber의 책임에 충실한 방향인 것 같다는 생각이 드네요

pkch93 replied:
ㅎㅎ 원시값을 객체로 포장해서 그 의미에 걸맞는 다양한 기능들을 제공할 수 있도록 만들어 줄 수 있죠. 써머가 생각한 1~45에 대한 검증을 LottoNumber에 둔다는 것도 원시값 포장을 사용했을때 이점이라고도 할 수 있겠네요 :)

hyewoncc commented on src/main/java/domain/PurchasedLotto.java:
```diff
@@ -0,0 +1,57 @@
+
+    private List<Lotto> lottos = new ArrayList<>();
+    private PrizeResult prizeResult = new PrizeResult();
+    private int inputMoney;
```
inputMoney도 포장되지 않은 원시값에 속할까요?
LottoNumber와 달리 개념적인 객체가 아니라 그냥 변수라고 생각해서 일단 하지 않았어요 
그리고 Lotto가 이미 LottoNumber가 모인 일급 컬렉션인데, 
이 것도 일급 컬렉션으로 처리해야 할지, 그러면 일급 컬렉션 안의 일급 컬렉션이 되는 것인데 괜찮은지 궁금합니다

pkch93 replied:
`List<Lotto>`를 하나의 객체로 감쌌을때 어떤 이점이 있을지가 먼저 선행되어야할 거 같아요. inputMoney도 마찬가지구요. inputMoney를 현재 어떻게 사용하고 있는지 보고 객체로 감싸 역할을 부여할지 등을 보면 좋을거 같아요. (제가 보기엔 inputMoney는 없어져도 될거 같기도 하네요)

hyewoncc commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,46 @@
+        Integer[] winningNumbersArray = {1, 2, 3, 10, 11, 12};
+        List<Integer> winningNumbers = Arrays.asList(winningNumbersArray);
+        WinningNumber winningLotto = new WinningNumber(winningNumbers);
+        winningLotto.addBonusNumber(45);
```
전략 패턴을 이용해 로또 번호를 조작하는 식으로 테스트를 작성했는데요, 
케이스를 충분히 늘려야 하는가? 아니면 논리적으로 한 케이스로 충분할 때 하나만 해도 되는가 궁금합니다

pkch93 replied:
테스트 케이스 작성에 대한 고민으로 보이네요.
위 부분은 제 생각으로는 각 요구사항을 구현한 로직은 반드시 테스트가 되어야한다고 생각해요. 1, 2, 3, 4, 5, 미당첨에 대해서 모두 테스트가 필요하다고 생각합니다.

hyewoncc commented on src/test/java/domain/WinningNumberTest.java:
```diff
@@ -0,0 +1,26 @@
+                .hasMessageContaining("보너스 번호가 당첨 번호와 중복됩니다.");
+    }
+
+}
```
앗 자동 포맷팅을 사용했는데 빈 줄이 없는 파일이 있네요 ㅠㅠ 수정하겠습니다

hyewoncc reviewed (commented):
다음 단계에서 어떻게 하면 원시값 포장의 이점을 살릴 수 있을 지가 현재로서는 가장 고민됩니다
이번 리뷰 잘부탁드려요 🙇

pkch93 commented on src/main/java/controller/LottoGameController.java:
```diff
@@ -0,0 +1,36 @@
+        OutputView.printEarningRate(purchasedLotto.calculateEarningRate());
+    }
+
+    private void initWinningNumber() {
```
initWinningNumber와 initBonusNumber도 InputView의 다른메서드들 처럼 InputView에서 제공해줄 수 있지 않을까요?

pkch93 commented on src/test/java/domain/PurchasedLottosTest.java:
```diff
@@ -0,0 +1,55 @@
+    void init() {
+        inputMoney = 1400;
+        lottos = new PurchasedLotto(inputMoney);
+        lottos.purchase(new CustomPurchaseStrategy(Arrays.asList(new Integer[]{1, 2, 3, 43, 44, 45})));
```
```suggestion
        lottos.purchase(new CustomPurchaseStrategy(Arrays.asList(1, 2, 3, 43, 44, 45)));
```
`Arrays.asList`를 쓰셨던 만큼 위와 같이 쓰실수도 있어요~ Java 9 부터는 `List.of`도 사용할 수 있어서 참고해주시면 될거 같아요. (`List.of` 사용을 더 권장합니다)

참고: https://stackoverflow.com/questions/46579074/what-is-the-difference-between-list-of-and-arrays-aslist

pkch93 commented on src/main/java/domain/PurchasedLotto.java:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    public void purchase(PurchaseStrategy purchaseStrategy) {
```
테스트를 위해서 public으로 여신 느낌이네요. 위 메서드가 없이도 테스트가 가능하도록 구조를 바꿔볼 수 있을거 같아요. 한번 고민해보시면 좋겠습니다~

hyewoncc replied:
며칠 고민해봤는데 현재로서는 대안이 떠오르지 않아서 더 진행하며 생각해보겠습니다 😭

pkch93 replied:
PurchasedLotto가 사라지고 LottoGame이 그 역할을 대체하는걸로 보이네요. 다만, LottoGame과 PurchasedLotto의 구조는 동일해보이네요 ㅎㅎ

아마 `RandomPurchaseStrategy`가 가진 랜덤 로직때문에 `purchase`를 public으로 열고 테스트용 PurchaseStrategy를 사용하신게 아닌가 싶어요.

현재 `purchase` 메서드가 게임에 사용될 lotto를 추가하는 로직이기 때문에 추후 유지보수를 하는 과정에서 잘못 사용이 될 수 있어보여요. 이런 경우를 아예 차단하기 위해서 LottoGame을 생성할 때 로또를 받아서 생성하면 어떨까요? 다음 단계 진행하시면서 같이 고민해보면 좋을거 같습니다 :)

pkch93 commented on src/main/java/domain/PurchasedLotto.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    public PrizeResult calculatePrizeResult(WinningNumber winningNumber) {
+        lottos.stream()
```
```suggestion
        lottos
```
`foreach`를 쓰신다면 `stream`은 삭제해도 무방해요

pkch93 commented on src/main/java/domain/PurchasedLotto.java:
```diff
@@ -0,0 +1,57 @@
+        long totalPrize = prizeResult.totalPrize();
+
+        float earningRate = (float) totalPrize / inputMoney;
+        return (float) (Math.floor(earningRate * 100) / 100.0);
```
double 대신 float를 사용하신 이유가 있을까요?

hyewoncc replied:
페어랑 예제를 보니 수익률의 경우 소수점 셋째자리에서 내림을 했길래 float로 충분하다 생각했어요
그런데 물어보신 이유가 있을 것 같아 찾아보니, 자바 실수 연산은 double 형을 기본으로 한다고 나와서 수정했습니다

pkch93 commented on src/main/java/domain/WinningNumber.java:
```diff
@@ -0,0 +1,34 @@
+        numbers.forEach(number -> winningNumbers.add(new LottoNumber(number)));
+    }
+
+    public void addBonusNumber(int number) {
```
생성자에서 보너스 숫자를 같이 받는건 어떨까요?

pkch93 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    private int matchedRegularNumbers(WinningNumber winningNumber) {
+        return winningNumber.getWinningNumbers().stream()
```
WinningNumber의 숫자를 가져와서 판단하고 있네요. WinningNumber에 얼마나 일치하는지를 판단할 수 있도록 책임을 더 주면 어떨까요?

pkch93 reviewed (changes requested):
안녕하세요 써머~! 이번 써머의 로또 미션 리뷰어를 맡게된 철시입니다 :)
저도 잘부탁드려요 ㅎㅎ

질문 주신 부분에 코멘트 드렸구요. 저도 몇몇 피드백 코멘트로 남겼습니다. 확인 부탁드릴께요~!

hyewoncc committed:
refactor: 로또 숫자 객체 이름 변경, 유효성 검사 역할 이양
LottoNumber -> LottoBall 로 이름 변경
숫자 1~45 범위 유효성 검사를 LottoBall이 하도록 변경
변경에 따른 테스트 추가와 삭제

hyewoncc committed:
refacotr: 불필요한 stream 제거

hyewoncc committed:
refactor: 메서드 명 오탈자 정정, 불필요한 stream 제거

hyewoncc committed:
refactor: 클래스 이름 변경
WinningNumber -> WinningLotto

hyewoncc committed:
refactor: 컬렉션 초기화 개선

hyewoncc committed:
refactor: WinningLotto의 생성 파라미터 추가, Prize 구하기 이양

hyewoncc committed:
style: 파일 끝 빈줄

hyewoncc committed:
feat: Lotto와 WinningLotto의 테스트 케이스 추가

hyewoncc committed:
refactor: 메서드 이름 개선

hyewoncc committed:
refactor: Lotto 유효성 검사를 Lotto로 이동
이동에 따른 테스트 이동과 케이스 추가 작성

hyewoncc committed:
fix: 유효하지 않은 우승 로또 입력 시 즉시 예외 발생

hyewoncc committed:
refactor: 클래스 이름 변경

hyewoncc committed:
refactor: inputMoney 인스턴스 필드 제거

hyewoncc committed:
컨트롤러의 인스턴스 필드 제거

hyewoncc committed:
refactor: 메서드의 반환형을 double로 변경

hyewoncc committed:
fix: 생성자 파라미터 변경에 따른 테스트 케이스 수정

hyewoncc committed:
refactor: 최소 금액 검증 책임을 LottoGame 으로 이동

hyewoncc committed:
refactor: 출력 포맷 적용 개선

hyewoncc committed:
refactor: 수익률 계산의 반환형을 float 에서 double 로 변경

hyewoncc committed:
style: 오탈자 정정

hyewoncc commented:
안녕하세요, 첼시! 남겨주신 코멘트 감사하게 읽었어요
객체의 책임이 잘못 나뉘어져 있거나 모호했던 부분을 바꾸는 걸 중점으로 진행했습니다
지난 미션에서 다른 크루들의 코드를 너무 일찍 보고 다녀서, 제가 스스로 생각하고 해답을 찾은 부분이 별로 없었던 것 같아 
이번 미션은 되도록 마지막 pr 전까지는 혼자 고민해보려 했는데, 답을 못 구한 부분이 있는 게 아쉽네요 😢
이번 리뷰도 잘부탁드려요!

pkch93 commented on src/test/java/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,31 @@
+                Arguments.of(new Lotto(List.of(10, 15, 20, 25, 30, 35)), Prize.FIRST));
+    }
+
+    @ParameterizedTest
```
`@ParameterizedTest`를 사용하셨네요 👍 

`MethodSource`의 메서드 - 본 테스트 순서로 정의를 하셨는데요. 위 순서에 대해서도 한번 고민해보시면 좋을거 같습니다.

저 같은 경우는 과정으로 이뤄지는 테스트인지 테스트 본체를 먼저 정의하고 해당 테스트에 필요한 데이터를 확인할 수 있도록 배치하는데요. 테스트가 어떻게 이뤄지는지고 주된 관심사이고 다음에 어떤 데이터들이 테스트되는지 보는 순서가 읽기 좋다고 생각해서에요.

이부분에 정답은 없긴해서 한번 고민해보면 좋을거 같아서 남깁니다 :)

pkch93 commented on src/main/java/utils/InputValidation.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class InputValidation {
```
inputValidation의 이름만 봐서는 값에 대한 검증을 담당하는 것으로 보이는데요. 현재 값을 변환하는 부분도 같이 담당하는 것으로 보이네요. Validation의 의미에 맞도록 검증만 담당하도록 만들면 어떨까 싶은데요. 한번 같이 고민해주시면 좋을거 같아요.

pkch93 commented on src/main/java/domain/PrizeResult.java:
```diff
@@ -0,0 +1,43 @@
+
+    public void updatePrizeResult(Prize winnerPrice) {
+        prizeResult.put(winnerPrice, prizeResult.get(winnerPrice) + 1);
+    }
```
PrizeResult가 로또 게임의 최종 결과를 담은 클래스로 보이는데요. updatePrizeResult가 꼭 필요한 메서드인지 의문이 들어요. 위 메서드를 없애볼 수는 없을지 고민해보면 좋을거 같아요.

pkch93 reviewed (approved):
써머 안녕하세요~
리팩토링을 통해 어느정도 개선이 된거 같아요 :) 이번 단계는 마무리 해도 될거 같아서 이만 머지할께요 ㅎㅎ
다음 단계에서 같이 고민해보면 좋을 것들도 같이 코멘트 남겼어요. 한번 고민해주시면 좋을거 같습니다 :)
2단계에서 뵐께요 👋
